### PR TITLE
Mdct 2005 - Add dropdown handling

### PR DESCRIPTION
**Pull request by gmrabian**
## Description
[MDCT-2005](https://qmacbis.atlassian.net/browse/MDCT-2005) Setup dropdown fields to follow our choice pattern, to hydrate, and to work in nested fields

### How to test
You will need to manually add a dropdown to the mcpar.json. You can do this anywhere you want! I suggest Section A: point of contact for a top level one 
```
{
  "id": "test dropdown",
  "type": "dropdown",
  "validation": "dropdown",
  "props": {
    "label": "test",
    "hint": "test",
    "options": [
      {
        "label": "string",
        "value": "test"
      },
      {
        "label": "string2",
        "value": "test2"
      }
    ]
  }
}
```

and Section B: state-level-indicators for a nested one! (under other, specify)

```
{
  "id": "test dropdown",
  "type": "dropdown",
  "validation": {
    "type": "dropdown",
    "nested": true,
    "parentFieldName": "state_encounterDataValidationEntity"
  },
  "props": {
    "label": "test",
    "hint": "test",
    "options": [
      {
        "label": "string",
        "value": "test"
      },
      {
        "label": "string2",
        "value": "test2"
      }
    ]
  }
}
```

## Code author checklist
- [ ] I have performed a self-review of my code
- [ ] I have added [thorough](https://bit.ly/3zPrxuZ) tests
- [ ] I have added analytics, if necessary
- [ ] I have updated the documentation, if necessary

## Reviewer checklist (two different people)
- [ ] I have done the deep review and verified the items in the above checklist are g2g
- [ ] I have done the lgtm review

### CODE REVIEW

File: services/app-api/utils/validation/schemaMap.ts [hunk 1]
```diff
@@ -122,7 +122,7 @@ export const endDate = (startDateField: string) =>
 
 // DROPDOWN
 export const dropdown = () =>
-  string().typeError(error.INVALID_GENERIC).required(error.REQUIRED_GENERIC);
+  object({ label: text(), value: text() }).required(error.REQUIRED_GENERIC);
```
Comment (gmrabian) [inline services/app-api/utils/validation/schemaMap.ts:125]: Devnotes: updated dropdown schema to be similar to how we handle choice fields. You'll see in this change in the ui-src/ schemas as well

File: services/app-api/utils/validation/schemaMap.ts [hunk 2]
```diff
@@ -157,6 +157,7 @@ export const nested = (
     mixed: number(),
     string: string(),
     date: date(),
+    object: object(),
```
Comment (gmrabian) [inline services/app-api/utils/validation/schemaMap.ts:160]: Devnotes: For handling dropdowns in nested fields. It automatically detects it from the generic `object` type

File: services/ui-src/src/components/fields/ChoiceListField.tsx [hunk 3]
```diff
@@ -91,6 +92,13 @@ export const ChoiceListField = ({
                 clearNestedValues(child.props.choices);
               }
               break;
+            case "dropdown":
+              form.setValue(
+                child.id,
+                { label: dropdownDefaultOptionText, value: "" },
+                { shouldValidate: true }
+              );
+              break;
```
Comment (gmrabian) [inline services/ui-src/src/components/fields/ChoiceListField.tsx:101]: Devnotes: allow dropdown reset when parent field changes

**Review by braxex — APPROVED** [review]
Approved with one question.

File: services/ui-src/src/components/fields/DropdownField.tsx [hunk 4]
```diff
@@ -77,14 +49,48 @@ export const DropdownField = ({
...
+      value: event.target.value,
+    };
+    setDisplayValue(selectedOption);
+    form.setValue(name, selectedOption, { shouldValidate: true });
+  };
+
   // prepare error message, hint, and classes
   const formErrorState = form?.formState?.errors;
-  const errorMessage = formErrorState?.[name]?.message;
+  const errorMessage = formErrorState?.[name]?.value.message;
```
[the remark this answers is not in the document]
Comment (gmrabian) [reply services/ui-src/src/components/fields/DropdownField.tsx:87]: I think we'd only need it if `value` might not exist on a dropdown field, but `value` is required in the Dropdown type. In the event there is no error, `[name]` will not exist and it'll be undefined anyway, so safe I believe.

[the remark this answers is not in the document]
Comment (braxex) [reply services/ui-src/src/components/fields/DropdownField.tsx:87]: Ah yeah, you're right. 👍

**Review by ntsummers1 — APPROVED** [review]

**Review by karla-vm — APPROVED** [review]

**Review by britt-mo — APPROVED** [review]

**Pull request merged**